consensus/istanbul/backend/engine_test.go

```diff
@@ -812,3 +821,184 @@ func TestSnapshot(t *testing.T) {
 		engine.Stop()
 	}
 }
+
+func TestGovernance_Votes(t *testing.T) {
+	type vote struct {
+		key   string
+		value interface{}
+	}
+	type governanceItem struct {
+		vote
+		appliedBlockNumber uint64 // if applied block number is 0, then it checks the item on current block
+	}
+	type testcase struct {
+		votes    []vote
+		expected []governanceItem
+	}
+
+	testcases := []testcase{
+		{
+			votes: []vote{
+				{"governance.governancemode", "none"},     // voted on block 1
+				{"istanbul.committeesize", uint64(4)},     // voted on block 2
+				{"governance.unitprice", uint64(2000000)}, // voted on block 3
+				{"reward.mintingamount", "96000000000"},   // voted on block 4
+				{"reward.ratio", "34/33/33"},              // voted on block 5
+				{"reward.useginicoeff", true},             // voted on block 6
+				{"reward.minimumstake", "5000000"},        // voted on block 7
+			},
+			expected: []governanceItem{
+				{vote{"governance.governancemode", "none"}, 6},
+				{vote{"istanbul.committeesize", uint64(4)}, 6},
+				{vote{"governance.unitprice", uint64(2000000)}, 9},
+				{vote{"reward.mintingamount", "96000000000"}, 9},
+				{vote{"reward.ratio", "34/33/33"}, 9},
+				{vote{"reward.useginicoeff", true}, 12},
+				{vote{"reward.minimumstake", "5000000"}, 12},
+				// check governance items on current block
+				{vote{"governance.governancemode", "none"}, 0},
+				{vote{"istanbul.committeesize", uint64(4)}, 0},
+				{vote{"governance.unitprice", uint64(2000000)}, 0},
+				{vote{"reward.mintingamount", "96000000000"}, 0},
+				{vote{"reward.ratio", "34/33/33"}, 0},
+				{vote{"reward.useginicoeff", true}, 0},
+				{vote{"reward.minimumstake", "5000000"}, 0},
+			},
+		},
+		{
+			votes: []vote{
+				{"governance.governancemode", "none"},   // voted on block 1
+				{"governance.governancemode", "single"}, // voted on block 2
+				{"governance.governancemode", "none"},   // voted on block 3
+				{"governance.governancemode", "single"}, // voted on block 4
+				{"governance.governancemode", "none"},   // voted on block 5
+				{"governance.governancemode", "single"}, // voted on block 6
+				{"governance.governancemode", "none"},   // voted on block 7
+				{"governance.governancemode", "single"}, // voted on block 8
+				{"governance.governancemode", "none"},   // voted on block 9
+			},
+			expected: []governanceItem{
+				{vote{"governance.governancemode", "single"}, 6},
+				{vote{"governance.governancemode", "none"}, 9},
+				{vote{"governance.governancemode", "single"}, 12},
+				{vote{"governance.governancemode", "none"}, 15},
```

Comment:
I can't understand this.
The last vote "none" was in block 9 and the epoch was 3. I think the vote is applied from block 12.

Reply:
My mistake. The vote is effective after 2 * epoch later, right? Then, your code makes sense.

Reply:
Yes, it is applied after 2 epochs.